eras/shelley/impl/src/Cardano/Ledger/Shelley/Core.hs

```diff
@@ -23,21 +27,50 @@ import Cardano.Ledger.Shelley.Delegation.Certificates (DCert)
 import Cardano.Ledger.Shelley.Era (ShelleyEra)
 import Cardano.Ledger.Shelley.PParams (Update)
 import Cardano.Ledger.Slot (SlotNo (..))
+import Cardano.Ledger.TreeDiff (ToExpr)
 import Control.DeepSeq (NFData)
+import Data.Default.Class (Default (..))
+import Data.Kind (Type)
 import Data.Map.Strict (Map)
 import Data.Maybe.Strict (StrictMaybe)
 import Data.Sequence.Strict (StrictSeq)
 import GHC.Generics (Generic)
 import Lens.Micro (Lens', SimpleGetter)
 import NoThunks.Class (NoThunks)
 
+class EraTallyState era where
+  type TallyState era :: Type
+  type TallyState era = ShelleyTallyState era
```

Comment:
I'm trying remember the rationale for calling this the **shelley** tally state. I think it was because it is first used introduced in the shelley ledger era, even though it will be `emptyTallyState` until Conway, yea?

Reply:
We're adding a `TallyState` to the `LedgerState` so we need some sort of a placeholder for pre-Conway eras. The `ShelleyLedgerState` is basically just a unit type so we can still construct a ledger state, but the "tally state" won't carry any information with it. I `guess` we could also use a unit instead, but it doesn't really make a difference.

Reply:
The motivation is by using `ShelleyTallyState era`, we can make `TallyState` invective. So you can change this and it will improve type inference:

```haskell
type TallyState era = (r :: Type) | r -> era
```

Reply:
Oh, that explains why I had to add `@era` everywhere..

Reply:
Ok, I added the injective type family annotation and got rid of some `@era`s